include/jemalloc/internal/sized_alloc_region.h

```diff
@@ -0,0 +1,144 @@
+#ifndef JEMALLOC_INTERNAL_SIZED_ALLOC_REGION_H
+#define JEMALLOC_INTERNAL_SIZED_ALLOC_REGION_H
+
+#include "jemalloc/internal/alloc_ctx.h"
+#include "jemalloc/internal/assert.h"
+#include "jemalloc/internal/atomic.h"
+#include "jemalloc/internal/pages.h"
+
+/*
+ * This file defines sized alloc regions.  These regions have portions of memory
+ * dedicated to a particular size class.  When allocations come out of these
+ * regions, we can quickly determine their size based only on their address and
+ * a little bit of global metadata, thus avoiding any rtree lookups or (if we're
+ * lucky) any other cache-unfriendly lookups at all.
+ *
+ * We orchestrate it so that zero-initialization is a valid state for the
+ * sized_alloc_region_t.  This lets us avoid branching on
+ * opt_sized_alloc_region.
+ */
+
+/*
+ * SIZED_ALLOC_REGION_LG_NUM_SCS:
+ *     The log of the number of size classes that can come out of the
+ *     sized-alloc region.
+ * SIZED_ALLOC_REGION_LG_SC_SIZE:
+ *     The log of the size of virtual address space that we allocate to each
+ *     size class within the sized-alloc region.
+ */
+#if LG_SIZEOF_PTR == 2
```

Comment:
It wouldn't be crazy. It needs to be a power of two so we can get by with just a shift. It would move the shift amount from being embdedded in the instruction into being a load + dynamic shift; not sure if we would notice.

These settings allow for 4GB of allocations from each size class (in 64 bit mode) though; this exceeds anything I've seen even from big services.